CM7/Core/Src/encoder.c

```diff
@@ -0,0 +1,51 @@
+#include "encoder.h"
+
+#include <stdlib.h>
+#include <assert.h>
+#include <math.h>
+
+#include "stm32h7xx.h"
+#include "stm32h7xx_hal_spi.h"
+#include "stm32h7xx_hal_tim.h"
+
+#define ENCODER_MAX_COUNT ((1 << 13) - 1)
+#define PULSES_PER_REV    (2048) /* from datasheet */
+
+encoder_t *encoder_init(SPI_HandleTypeDef *spi_handle, TIM_HandleTypeDef *timer_handle)
+{
+    // make sure the handle exists
+    assert(spi_handle);
```

Comment:
maybe assert timer handle too?

Reply:
oh ya you're so right, I'll make that change now